### ISSUE TITLE
The program only worked once, but never again..

### ISSUE THREAD

**Issue by Madridi7** [soarqin/finalhe#22]
First of all, thanks for this program. It's fantastic. I used 1.4 last night, and it worked without any problem, and is much faster and cleaner than auto-hencore.

Today, I don't know what changed. I tried to use it for another vita, and it no longer works correctly. It can detect and fetch my AID fine. It proceeds to download and unpack all necessary files just fine.

The problem starts happening when it says "Createing psvimg's" (Typo btw, the correct spelling is creating).. It goes from 0 to 90%, then quickly finishes.

Upon checking \DATA\h-encore\PCSG90096\app, I see that app.psvimg is only 32 bytes.
Then checking \DATA\h-encore\PCSG90096\license, I once again see License.psvimg only being 32 bytes.

1.5 changed nothing. Same behavior. But interestingly enough, I redownloaded 1.4, and now the folder layout is different. With 1.5, the hencore folder only has PCSG90096 inside of it, with the other 5 folders inside it. But with 1.4, it has PCSG90096 + 4 of the 5 folders there (sce_sys) is not there. Upon checking these folders, it seem like they are unpacked PCSG90096 game that was is not getting cleaned properly. Also, the creating psvimg's screen finishes instanly.

I went back and tried auto-hencore, that worked fine, so it doesnt seem like a driver issue or anything on my side. Also, since it worked yesterday, I dont think I have anything wrong on my PC.

For reference, I am using a windows 7 64bit PC

**Comment by soarqin** [comment]
this should be caused by failure of fetching backup key from cma.henkaku.xyz, and the bad key is cached in C:\Users\<your account>\AppData\Roaming\soarqin\FinalHE.ini(v1.5) or C:\Users\<your account>\AppData\Roaming\FinalHE\FinalHE.ini(v1.4)
Try remove this file and re-run the application.

**Comment by Madridi7** [comment]
Just to follow up, I tried yet another vita. Same account I just tried, but never signed to that account (entered only through the first initialization of the system). Therefore, instead of seeing my PSN name, it sees user328. So it says: Connected to PS Vita @368 [user328](AID NUMBER)

The result: Same behavior as earlier, 32byte files. So I tried deleting the ini again, and retried. Same result.

Do I have to sign to PSN for this to work? This is again something that I auto-hencore did without having to sign to PSN.

**Comment by soarqin** [comment]
This seems like a bug in my code, I am going to check this bug asap, thanks for report.

**Comment by Madridi7** [comment]
Happy to help. Let me know if there is anything I can do to help.

**Comment by soarqin** [comment]
@Madridi7 One more question: is that backup key in FinalHE.ini 64 characters long?

**Comment by Madridi7** [comment]
Yes, from what I see, it shows:
AID=64 characters

**Comment by soarqin** [comment]
so this can be something wrong with the win7 process, I added an extra check to verify length of backup key, please try this to see if the problem is fixed: [FinalHE_test1.zip](https://github.com/soarqin/finalhe/files/2246571/FinalHE_test1.zip)

**Comment by Madridi7** [comment]
I fail to see the reason? across all ini files, the backupkeys were identical. It's the [General] section that was changing. So I'm not sure why even deleting it worked at some point if [General] is irrelevant.

That said, the test build changes nothing. Same behavior

**Comment by soarqin** [comment]
UUID in General is just a random generated thing for WiFi connection use, it totally useless for USB connection.

**Comment by soarqin** [comment]
if not cleaning FinalHE.ini. can you see 2 lines in [BackupKeys] after 2nd device connected?

**Comment by Madridi7** [comment]
I see, so in other words the only relevant part is the [backupkeys], which is identical with the successful and failed install.

No, I wouldve thought that my original ini, before you asked me to delete it, would have 2 lines, one for my friend's AID and one for mine, but it only had mine.

Now to further complicate things. I cleared all my ini files. I went back to my vita that just worked earlier (the one with logged in PSN), and the process no longer works! 32 byte files again..

**Comment by soarqin** [comment]
I see, I will get another Vita from my friend for test on Win7 VM tomorrow, hope to find what is wrong with it

**Comment by Madridi7** [comment]
I found finalHE.ini in another folder called "unknown organization". 1.3 crashed when I attempted it. I assume it's from that version. The file did not have backupkeys line. I deleted that entire folder and tried the vita with the logged psn account, and it worked again this time. I'll go ahead and try with the other vita now and report back

**Comment by soarqin** [comment]
yes, 1.3 release does not use backup keys in FinalHE.ini, and I forgot to set Organization so that it is named "unknown organization".

I think logged psn account is not really need, although I have not tested it.

**Comment by Madridi7** [comment]
Ok, I need to do more tests to confirm the following is actually the case. But with the non logged vita, it didnt work.

So once it didnt work, I didnt delete the ini file. I deleted the whole soarqin folder this time. To my surprise, it worked!

I will try several tries to confirm its the case, or is it just seemingly random

**Comment by soarqin** [comment]
if it is just a randomly occured issue, i should assume that the point might be the bad support of _pipe on win7, I will try to rewrite psvimg creation functions to use mutex/sema later

**Comment by Madridi7** [comment]
Seems random .. It now keeps failing again :(
Not sure what else I could try. If you can get your hands on a second vita, maybe you can try going back and forth between them and see if it works everytime

**Comment by Madridi7** [comment]
Not sure if this is of any help, but I also noticed that the "trim rom" functionality is entirely skipped when it fails. It goes directly from unpacking to creating psvimg's (32 bytes)

**Comment by soarqin** [comment]
Totally removed threading and pipe from psvimg creation, hope this fixes your issue: 
[FinalHE.zip](https://github.com/soarqin/finalhe/files/2252148/FinalHE.zip)

**Comment by Madridi7** [comment]
Unfortunately, it did not. Failed on the first try, with the soarqin folder deleted before trying

**Comment by soarqin** [comment]
I have tested this in my clean win7 in VM and everything worked fine, I think there might be some file lock in your OS, such like anti-virus scanning or additional I/O driver with monitoring function, I will compile a client with console logging window for you to diagnostic asap

**Comment by soarqin** [comment]
[FinalHE_test2.zip](https://github.com/soarqin/finalhe/files/2252231/FinalHE_test2.zip)
here it is, check popup console window during psvimg packing for errors and report back.

**Comment by Madridi7** [comment]
No need to test. I just found the problem. You are right, it is my anti-virus! I just disabled it and tried 1.5 again, it worked 5 times in a row. I am using Kaspersky antivirus

I also noticed it is MUCH faster than when my antivirus was on. The progress bars goes by as a rocket. Unpacking used to take around 10 seconds, now its only like, 2-3s .. Trimming package is also done in less than 1s. It used to take maybe 5s

Now that you mention antivirus, 1.5 finalhe.exe in particular was detected as a trojan when I first used it. I was sure it was just a false positive, so I had to exclude it to get it to run. But apparently messing up with the stuff happening in data folder?

Would you still like me to test the latest build or is this info sufficient? Any way around it or do I have to disable antivirus everytime I want to use it?

**Comment by soarqin** [comment]
This turns it hard to fix, you can add the whole directory to whitelist, or try set antivirus real-time monitor to scan only executable files that may reduce time penalty and thus issue. But anyhow I will try to find a solution if possible.

**Comment by Madridi7** [comment]
Yeah cool, I can do that as a workaround and it’s not a big deal anyway. It would be great if you can get it to work but I don’t think it’s possible since I assume everything happening is from antivirus checking through the unpacked pkg. I was just going nuts to why it would just randomly work.

Thanks a lot for your help and patience on this issue. Good eye thinking it might be my antivirus.

Not sure if I should keep this issue open or close it. Feel free to close it if it’s no longer relevant
